File: src/main/java/me/snowmii/streamline/MotionRequest.java
package me.snowmii.streamline;

import java.util.Arrays;
import java.util.Objects;

/**
 * One motion pass, in the units the flat native ABI takes them.
 *
 * <p>{@link #reprojection} is the 16 column-major floats of
 * {@code DlssFrameMotion.reprojection}, which maps a jittered clip position to the previous
 * frame's unjittered one. The destination is the native motion image, which the bridge owns,
 * so it never appears here.
 *
 * <p>{@link #renderDimensions} is stamped by the session adapter rather than supplied here,
 * for the same reason as in {@link EvaluationRequest}.
 */
public record MotionRequest(
	/** The caller's shared Vulkan command buffer the pass is recorded on. */
	long commandBuffer,
	/** The engine's render-sized depth image. */
	ImageBinding depth,
	/** 16 column-major reprojection floats mapping this frame's clip to the previous one's. */
	float[] reprojection,
	/** Stamped by the session adapter; see the class comment. */
	Dimensions renderDimensions
) {
	public MotionRequest {
		Objects.requireNonNull(depth, "depth");
		Objects.requireNonNull(reprojection, "reprojection");
	}

	/**
	 * Compares the reprojection payload, not the array identity the generated {@code equals}
	 * would compare - the Kotlin data class this record replaces did the same.
	 */
	@Override
	public boolean equals(Object other) {
		return other instanceof MotionRequest request
			&& request.commandBuffer == commandBuffer
			&& request.depth.equals(depth)
			&& Arrays.equals(request.reprojection, reprojection)
			&& Objects.equals(request.renderDimensions, renderDimensions);
	}

	@Override
	public int hashCode() {
		int result = Long.hashCode(commandBuffer);
		result = 31 * result + depth.hashCode();
		result = 31 * result + Arrays.hashCode(reprojection);
		result = 31 * result + Objects.hashCode(renderDimensions);
		return result;
	}
}

File: src/main/kotlin/me/snowmii/dlss/bridge/MotionRequest.kt
package me.snowmii.dlss.bridge

/**
 * One motion pass, in the units the flat native ABI takes them.
 *
 * [reprojection] is the 16 column-major floats of `DlssFrameMotion.reprojection`, which maps a
 * jittered clip position to the previous frame's unjittered one. The destination is the native
 * motion image, which the bridge owns, so it never appears here.
 *
 * [renderDimensions] is stamped by [me.snowmii.dlss.session.LifecycleAdapter] rather than
 * supplied here, for the same reason as in [EvaluationRequest].
 */
data class MotionRequest(
	val commandBuffer: Long = 0,
	val depth: ImageBinding = ImageBinding(0, 0, 0),
	val reprojection: FloatArray = FloatArray(16),
	/** Stamped by the adapter; see the class comment. */
	val renderDimensions: DlssDimensions? = null,
) {
	// FloatArray has identity equals, which a data class would silently inherit.
	override fun equals(other: Any?): Boolean {
		if (this === other) return true
		if (other !is MotionRequest) return false
		return commandBuffer == other.commandBuffer &&
			depth == other.depth &&
			reprojection.contentEquals(other.reprojection) &&
			renderDimensions == other.renderDimensions
	}

	override fun hashCode(): Int {
		var result = commandBuffer.hashCode()
		result = 31 * result + depth.hashCode()
		result = 31 * result + reprojection.contentHashCode()
		result = 31 * result + renderDimensions.hashCode()
		return result
	}
}
